実装機能：年月でのソート (PORTAL-53)

BSS-OnumaTakuto opened the pull request:
# Jira Ticket
## 1．年月でのソート
https://bss-portal.atlassian.net/browse/PORTAL-53​
​
# やった事
## 1．年月でのソート
**notifications.sql** に格納されているデータを「年月と重要ボタン」で検索した内容を元に取得し、表示出来るようにしました。​
​
## なぜやるのか
## 1．年月でのソート
​現状年月での絞り込み機能がなく、一覧画面で全てのお知らせが表示されている状態のため、今後お知らせが増えてきたときにすぐ検索できるように。
​
# 動作確認
## 1．年月でのソート
​windowsのEclipseで実行し、挙動確認を行いました。

【お知らせ一覧画面】年月ソートのみ (通常のお知らせのみ表示)
<img width="960" alt="スクリーンショット 2023-11-02 141900" src="https://github.com/miaochshBss/ShainKintaiKanriBss/assets/147013917/327be0e9-abdd-4276-958d-74572b3bc90f">

【お知らせ一覧画面】年月ソート＋重要ソート (全てのお知らせを表示)
<img width="960" alt="スクリーンショット 2023-11-02 141915" src="https://github.com/miaochshBss/ShainKintaiKanriBss/assets/147013917/68e1c00e-643b-471f-9a5b-27b67ac6972c">

​
# Refs (レビューにあたって参考にすべき情報）(Optional)
​今まで全てのお知らせを取得する場合、**NotificationsInfoMapper.xml**の 「**id = "getNotificationsListInfoByUserName"**」を使用していましたが、
何を取得するための変数なのか名称でわかりづらいため、**NotificationsInfoMapper.xml**に「**id="getAllNotifications"**」を追加し、ソート前のお知らせ一覧画面はこちらから取得を行っています。
また、**NotificationsInfoMapper.xml**の 「**id = "getNotificationsListInfoByUserName"**」は、**NotificationsController.java**で使用していますが、**NotificationsController.java**の修正は今回の作業範囲から外れるため、xmlのid は削除せず残してあります。

お忙しいところ恐縮ですが、レビューの方よろしくお願いいたします。


# 追加対応 (2023/11/06)
## 1．年月でのソート
Figmaのモックでは年月ソートがプルダウンになっているが、テキスト入力で実装してしまいました。
テキスト入力だとヒューマンエラーの原因になるため、プルダウンメニューに修正し、プルダウンメニューの項目 「YYYYMM」はお知らせデータの重複を削除し、降順に並び替えたものを表示しています。

【お知らせ一覧画面】プルダウンメニュー展開
<img width="960" alt="スクリーンショット 2023-11-06 172105" src="https://github.com/miaochshBss/ShainKintaiKanriBss/assets/147013917/7e7c4d0b-ed1c-419d-bd7a-167f3a3b35ab">

【お知らせ一覧画面】年月ソートのみ (通常のお知らせのみ表示)
<img width="960" alt="スクリーンショット 2023-11-06 182704" src="https://github.com/miaochshBss/ShainKintaiKanriBss/assets/147013917/38c9e612-7fd2-4922-8139-9031ecc4a151">

【お知らせ一覧画面】年月ソート＋重要ソート (全てのお知らせを表示)
<img width="960" alt="スクリーンショット 2023-11-06 172126" src="https://github.com/miaochshBss/ShainKintaiKanriBss/assets/147013917/f7a07a49-dc2b-499e-9605-08b04fc183c3">

BSS-OnumaTakuto committed:
年月日でのソート (PORTAL-53)

BSS-OnumaTakuto committed:
機能要件の考え方を間違えていたので、修正しました。

bss-nagasawa commented on ShainKintaiKanri/src/main/java/jp/co/bss/kintai/controller/NotificationsSearchController.java:
```diff
@@ -20,26 +20,38 @@ public class NotificationsSearchController {
-        model.addAttribute("notificationsInfoData", notificationsData);
-        return "notificationsSearch";
-    }
+//	// 全てのお知らせ一覧を表示
```
不要なコメントアウトは削除してください

BSS-OnumaTakuto replied:
コメントアウトの削除を行いました。

bss-nagasawa commented on ShainKintaiKanri/src/main/java/jp/co/bss/kintai/controller/NotificationsSearchController.java:
```diff
@@ -20,26 +20,38 @@ public class NotificationsSearchController {
         List<NotificationsInfo> notificationsData;
-        if (status != null && status.equals("1")) {
-            notificationsData = notificationsService.getNormalNotifications();
+        if (yearMonth != null) {
```
後々、考えることになりますが、ネストの深さを考えた実装が必要にはなるので、タグ、フリーワードの複合検索になった際に、ソースの可読性やメンテナンス性にかかわるのでネスト制限３以下で実装できるように頭の片隅に置いておいてください。
今はこのコメントを確認するだけでいいです。

bss-nagasawa commented on ShainKintaiKanri/src/main/java/jp/co/bss/kintai/mapper/NotificationsInfoMapper.xml:
```diff
@@ -5,8 +5,22 @@
+		AND status = '1'; 
+	</select>
+	
+</mapper>
```
ファイルの最終行は改行で終わってください。
全ファイル共通のルールでお願いします。

BSS-OnumaTakuto replied:
承知いたしました。

bss-nagasawa commented on ShainKintaiKanri/src/main/resources/templates/notificationsSearch.html:
```diff
@@ -149,6 +149,7 @@
                     <label class="toggle_label"></label>
                     <span class="content_pos">重要</span>
                 </div>
+                <input id="searchDate" class="date-input" type='text' placeholder="yyyy / MM" name="yearMonth">
```
<img width="514" alt="image" src="https://github.com/miaochshBss/ShainKintaiKanriBss/assets/147142110/7da610c4-fd9c-4838-beb4-16408969feaa">
プルダウンでの実装でお願いします

bss-nagasawa replied:
DatePickerなどの導入検討をしてみてください。
https://xiyuan.jp/javascript/275/

BSS-OnumaTakuto replied:
年月ソートをプルダウンに修正いたしました。
プルダウン内の年月は**notifications.sql**の**creation_date**から抽出しており、並び順は降順としています。

bss-nagasawa commented on ShainKintaiKanri/src/main/resources/templates/notificationsSearch.html:
```diff
@@ -168,17 +169,27 @@
+            	url += 'status=1';
+        	}
+        
+        	if (date) {
```
現状だとurlパラメーターがYYYY/MMなのでYYYYMMでお願いします。

BSS-OnumaTakuto replied:
urlパラメータが**YYYYMM**と表示されるように修正しました。

bss-nagasawa commented on ShainKintaiKanri/src/main/resources/templates/notificationsSearch.html:
```diff
@@ -168,17 +169,27 @@
+        	}
+        
+        	if (date) {
+        		if (checkBox.checked){
```
このif文の意味はどういった意味ですか？
if部内で両方が同じ処理なのでif文は不要では？
さらに条件がcheckBox.checkedになっている意味も分かりません。

BSS-OnumaTakuto replied:
if (checkBox.checked) の記載は不要でした。
よく理解しないまま実装してしまったのが原因です。
以後、実装して動くことだけ確認するのではなく、コード理解を深めるよう努力します。
コードを見直し、内容を修正いたしました。

BSS-OnumaTakuto committed:
頂いたコメントの内容で修正しました。

年月検索をテキスト入力にしていましたが、プルダウンに変更し、プルダウン内の年月はデータから抽出したものを使用しています。

BSS-OnumaTakuto commented on ShainKintaiKanri/src/main/resources/templates/notificationsSearch.html:
2023/11/06
コード修正中に見やすくするため、タグやコード位置をフォーマットしてしまい、コードの大半が変更箇所になってしまいました。
もし現場参画中にソースをフォーマットしてしまい、変更箇所が変更した内容以上に出てしまった場合は、どのように対処すれば良いのでしょうか。

1．フォーマットはしてはいけない。
2．プッシュ前の確認で、差分が全てのコードに出ていたら、フォーマット前まで戻して、再度コード修正する。
など、対処方法をご教授いただきたいです。

差分が全て出てしまった状態でプッシュしてしまい、申し訳ございません。

BSS-OnumaTakuto replied:
2023/11/07
ブランチ「feature-div-05-PORTAL-52」の **notificationsSearch.html** をコピーし、再度修正作業を行い、今回の作業で生じた差分を確認していただけるように変更しました。

BSS-OnumaTakuto committed:
notificationsSearch.htmlを修正

フォーマットしたものをプッシュしてしまい、ほぼ全てのコードで差分がでてしまったので、PORTAL-52のコードをコピーし、再度コード修正を行い、差分の確認が出来るようにしました。
